In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# creating a path to load data fames to python 
path = r"C:\Users\pshan\Downloads\InstaCart Basket Analysis"

In [3]:
# loading orders_products_combined dataset into jupytor
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

In [4]:
# slicing a part of data frame 
df = df_ords_prods_merge[:1000000]

In [5]:
df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both

In [6]:
# a function to flag different price categories
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [7]:
# calling the funtion
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\pshan\AppData\Local\Temp\ipykernel_23208\8094453.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [8]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

### # Use loc() method on df subset to apply price range function.

In [9]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\pshan\AppData\Local\Temp\ipykernel_23208\1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [10]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 


In [11]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [12]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

### Running on entire data frame

In [13]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [14]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [15]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [17]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21865987
Low-range product     10126321
High-range product      412551
Name: count, dtype: int64

In [43]:
# checking for the price_range_loc column in df_ords_prods_merge head 
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day         busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy days       Average orders   
1  Regularly busy      Least busy days       Average orders   
2  Regularly busy      Least busy days           Most orders  
3      Least busy      Least busy days       Average orders   
4      Least busy      Least busy days           Most orders

### Busiest day of week

In [18]:
# which day most orders take place
df_ords_prods_merge['orders_day_of_week'].value_counts()

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

#### 0 stands for saturdays which means saturday is the busiest day for instacart followed by sunday and wednesday is the slowest day of week

In [20]:
# creating an empty list 'result' to record entry
result = []

# the for loop will go through each value in orders_day_of_week and
# compare its value with what you know are the busiest and slowest days, and assign it the corresponding string value.

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [22]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [23]:
#  create a new column within  ords_prods_merge dataframe and set it equal to result
df_ords_prods_merge['busiest_day'] = result

In [24]:
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

### finding two busiest days and 2 slowest days of week

In [25]:
df_ords_prods_merge['orders_day_of_week'].value_counts()

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

#### 0 and 1 are the busiest days and 3,4 are the slowest days at insta cart

In [32]:
# creating an empty list 'result' to record entry
result_2 = []

# the for loop will go through each value in orders_day_of_week and
# compare its value with the busiest and slowest days, and assign it the corresponding string value.

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value == 1:
    result_2.append("Busiest days")
  elif value == 4 or value == 3:
    result_2.append("Least busy days")
  else:
    result_2.append("Regularly busy days")

In [33]:
result_2

['Regularly busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Regularly busy days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Least busy days',
 'Busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Busiest days',
 'Busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Regularly busy days',
 'Least busy days',
 'Regularly busy days',
 'Least busy days',
 'Busiest days',
 'Busiest days',
 'Regularly busy days',
 'Least busy days',
 'Least busy days',
 'Regularly busy days',
 'Regularly busy days',
 'Busiest days',
 'Busiest days',
 'Regularly busy days',
 'Busiest day

In [34]:
#  create a new column within  ords_prods_merge dataframe and set it equal to result_2
df_ords_prods_merge['busiest_days'] = result_2

In [35]:
df_ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy days    12916111
Busiest days           11864412
Least busy days         7624336
Name: count, dtype: int64

In [39]:
# checking head of df to check the new added column
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day         busiest_days  
0  Regularly busy  Regularly busy days  
1  Regularly busy      Least busy days  
2  Regularly busy      Least busy days  
3      Least busy      Least busy days  
4      Least busy      Least busy days

### checking the accuracy of the column
#### as 0 and 1 makes the busiest days category adding up them makes 11864412 which is accurate
#### as 3 and 4 makes the least busy days category adding up them makes 7624336 which is accurate

In [37]:
df_ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

#### as they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders the hours should be devided in three segments [10, 11, 14, 15, 13, 12, 16, 9] as the Most orders, [17, 8, 18, 19, 20, 7, 21, 22] as the avergae orders and [23, 6, 0, 1, 5, 2, 4, 3] as fewest orders 

In [38]:
# creating an empty list 'result' to record entry
result_3 = []

# the for loop will go through each value in order_hour_of_day and
# compare its value with the most orders , average orders and fewest order and assign it the corresponding string value.

for value in df_ords_prods_merge["order_hour_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9] :
    result_3.append("Most orders")
  elif value in [17, 8, 18, 19, 20, 7, 21, 22]:
    result_3.append("Average orders ")
  else:
    result_3.append("Fewest orders")

In [40]:
result_3

['Average orders ',
 'Average orders ',
 'Most orders',
 'Average orders ',
 'Most orders',
 'Average orders ',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders ',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders ',
 'Most orders',
 'Average orders ',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders ',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders ',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders ',
 'Most orders',
 'Most orders',


In [41]:
# addign a new column in the df and assign the result_3 to better use the new column
df_ords_prods_merge['busiest_period_of_day'] = result_3

In [42]:
# Printing frequencyy of the column
df_ords_prods_merge['busiest_period_of_day'].value_counts( dropna = False)

busiest_period_of_day
Most orders        21118071
Average orders      9997651
Fewest orders       1289137
Name: count, dtype: int64

### checking the accuracy of the column
#### as [10, 11, 14, 15, 13, 12, 16, 9] makes the busiest hour category adding up them makes 21118071 which is accurate


### exporting the updating data frame

In [44]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_derived.pkl'))